node/pod.go

```diff
@@ -185,6 +185,9 @@ func (pc *PodController) updatePodStatus(ctx context.Context, podFromKubernetes
 	kPod.Lock()
 	podFromProvider := kPod.lastPodStatusReceivedFromProvider.DeepCopy()
 	kPod.Unlock()
+	if cmp.Equal(podFromKubernetes.Status, podFromProvider.Status) {
```

Comment:
I think we keep track of the last pod we received from provider. Can we do the deduping there? That we look at the last pod from we received from the provider, and if it wasn't the same, then we dedupe? Otherwise, this mechanism can lead to "false updates" and "missed updates" due to lag.

Also, I think this behaviour should be behind a config, because right now, firing away an update is a way to force a pod status update from VK -- which a provider may want to do if something interferes with the write to pod status.

Reply:
IMO, `lastPodStatusReceivedFromProvider` seems not like a good design. As you mentioned, that may cause `false updates` and `false updates`.  Why not we change it to a `map`, not just a variable.

Reply:
https://github.com/virtual-kubelet/virtual-kubelet/blob/master/node/pod.go#L180
We have checked here, so it would not directly lead to "false updates" or "missed updates". This commit just add one more check before updating status to K8S.

@cpuguy83  wdyt?

Reply:
I think that without some level of checking here, you can get into a situation where you can miss updates. 

For example, if you have a pod which has a start, and a container fails, and it's restarted? You could fail to send the successful start because you haven't received the update from API server indicating it's failed?

Reply:
If restart, pod running time would change. Other messages in `status.conditions` would also change. What's more, if pods' status has not changed, why we need update the status to apiserver.

As you said, `start- fail - restart`, update status or not would not change the result.

Reply:
Are the timestamps generated client side, or server side? I think client-side?

I think that although Kubernetes is a line-level system, people could be depending on this behaviour. 

I think it makes much more sense to compare it to the last pod we "sent" in the NotifyPods callback if we really want to dedupe. That way the behaviour is not breaking.

Reply:
I still think we should compare it here. If status has not changed, updating it to api-server is meaningless. And this issue has leaded to the rate limit of client-go https://github.com/virtual-kubelet/virtual-kubelet/issues/822

Reply:
What's your reasoning against suppressing duplicate pod status updates in the NotifyPods callback?

Reply:
No, I am not against this. I think we'd add check both in the `NotifyPods` and `UpdatePodStatus`

Reply:
Do you want to add it to NotifyPods? And then we can hash this one out later?

Reply:
> Do you want to add it to NotifyPods? And then we can hash this one out later?

Sure, done

Reply:
Do you want to do the pod update dedupe in a different PR, or drop this dedupe behaviour in this PR to get it through?

Reply:
In regards to:
> IMO, lastPodStatusReceivedFromProvider seems not like a good design. As you mentioned, that may cause false updates and false updates. Why not we change it to a map, not just a variable.

I think that there are two reasons behind this design (and I will not be one to defend it as being perfect):

1. We want thee ability to perform critical code operations on the k8s pod, and the provider pod under the same lock.
2. From a performance perspective, having to hold a *global* lock while doing these (relatively) simple operations gets slow quick.

Reply:
> Do you want to do the pod update dedupe in a different PR, or drop this dedupe behaviour in this PR to get it through?

Thanks, I will move `pod update dedupe` to a different PR

Reply:
> In regards to:
> 
> > IMO, lastPodStatusReceivedFromProvider seems not like a good design. As you mentioned, that may cause false updates and false updates. Why not we change it to a map, not just a variable.
> 
> I think that there are two reasons behind this design (and I will not be one to defend it as being perfect):
> 
> 1. We want thee ability to perform critical code operations on the k8s pod, and the provider pod under the same lock.
> 2. From a performance perspective, having to hold a _global_ lock while doing these (relatively) simple operations gets slow quick.

Do we have a plan to change this logic, e.g. a map `map[pod.UID]pod` to make it possible to process sync pod status in parallel.